File: src/main/java/org/jukeboxmc/entity/item/EntityItem.java
package org.jukeboxmc.entity.item;

import com.nukkitx.nbt.NbtMap;
import com.nukkitx.nbt.NbtMapBuilder;
import com.nukkitx.protocol.bedrock.packet.AddItemEntityPacket;
import com.nukkitx.protocol.bedrock.packet.TakeItemEntityPacket;
import org.jukeboxmc.Server;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.entity.Entity;
import org.jukeboxmc.entity.EntityType;
import org.jukeboxmc.event.player.PlayerPickupItemEvent;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.item.ItemDiamondSword;
import org.jukeboxmc.item.ItemType;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;

import java.util.concurrent.TimeUnit;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class EntityItem extends Entity {

    private Item item;
    private long pickupDelay;
    private boolean isReset;
    private Player playerHasThrown;

    @Override
    public void update( long currentTick ) {
        super.update( currentTick );

        if ( this.isClosed() ) {
            return;
        }
        if ( !this.isImmobile() ) {
            if ( !this.isOnGround() ) {
                this.velocity = this.velocity.subtract( 0, this.getGravity(), 0 );
            }

            this.checkObstruction( this.location.getX(), this.location.getY(), this.location.getZ() );
            this.move( this.velocity );

            float friction = 1 - this.getDrag();

            if ( this.onGround && ( Math.abs( this.velocity.getX() ) > 0.00001 || Math.abs( this.velocity.getZ() ) > 0.00001 ) ) {
                friction *= 0.6f;
            }

            this.velocity = this.velocity.multiply( friction, 1 - this.getDrag(), friction );

            if ( this.onGround ) {
                this.velocity = this.velocity.multiply( 0, -0.5f, 0 );
            }

            this.updateAbsoluteMovement();
        }

        if ( this.isCollided && !this.isReset && this.velocity.squaredLength() < 0.01f ) {
            this.setVelocity( Vector.zero() );
            this.isReset = true;
        }

        if ( this.age >= TimeUnit.MINUTES.toMillis( 5 ) / 50 ) {
            this.close();
        }
    }

    @Override
    public String getName() {
        return "EntityItem";
    }

    @Override
    public float getWidth() {
        return 0.25f;
    }

    @Override
    public float getHeight() {
        return 0.25f;
    }

    @Override
    public float getGravity() {
        return 0.04f;
    }

    @Override
    public EntityType getEntityType() {
        return EntityType.ITEM;
    }

    @Override
    public void saveEntity( NbtMapBuilder builder ) {
        super.saveEntity( builder );
        NbtMapBuilder itemBuilder = NbtMap.builder();
        itemBuilder.putString( "Name", this.getItem().getIdentifier() );
        itemBuilder.putShort( "Damage", (short) this.getItem().getMeta() );
        itemBuilder.putByte( "Count", (byte) this.getItem().getAmount() );
        builder.putCompound( "Item", itemBuilder.build() );

       if ( !this.item.getBlock().getBlockStates().isEmpty() ) {
           NbtMap blockStates = this.item.getBlock().getBlockStates();
           builder.putCompound( "Block", blockStates );
       }
    }

    @Override
    public void loadEntity( NbtMap compound ) {
        super.loadEntity( compound );

        NbtMap itemCompound = compound.getCompound( "Item" );
        String identifier = itemCompound.getString( "Name" );
        short damage = itemCompound.getShort( "Damage" );
        byte amount = itemCompound.getByte( "Count" );
        this.item = ItemType.get( identifier ).setMeta( damage ).setAmount( amount ).setNBT( itemCompound );
        this.setPickupDelay( 1, TimeUnit.SECONDS );
        this.setPlayerHasThrown( null );

        if ( compound.containsKey( "Block" ) ) {
            NbtMap blockCompound = compound.getCompound( "Block" );
            int blockRuntimeId = Block.STATES.get( identifier ).get( blockCompound );
            this.item.setBlockRuntimeId( blockRuntimeId );
        }
    }

    @Override
    public AddItemEntityPacket createSpawnPacket() {
        AddItemEntityPacket addItemEntityPacket = new AddItemEntityPacket();
        addItemEntityPacket.setRuntimeEntityId( this.entityId );
        addItemEntityPacket.setUniqueEntityId( this.entityId );
        addItemEntityPacket.setItemInHand( this.item.toNetwork() );
        addItemEntityPacket.setPosition( this.location.toVector3f() );
        addItemEntityPacket.setMotion( this.velocity.toVector3f() );
        addItemEntityPacket.getMetadata().putAll( this.metadata.getEntityDataMap() );
        return addItemEntityPacket;
    }

    @Override
    public void onCollideWithPlayer( Player player ) {
        if ( this.item == null ) return;
        if ( Server.getInstance().getCurrentTick() > this.pickupDelay && !this.isClosed() && !player.isDead() ) {
            PlayerPickupItemEvent playerPickupItemEvent = new PlayerPickupItemEvent( player, this.item );
            Server.getInstance().getPluginManager().callEvent( playerPickupItemEvent );
            if ( playerPickupItemEvent.isCancelled() || !player.getInventory().canAddItem( this.item )) {
                return;
            }

            TakeItemEntityPacket takeItemEntityPacket = new TakeItemEntityPacket();
            takeItemEntityPacket.setRuntimeEntityId( player.getEntityId() );
            takeItemEntityPacket.setItemRuntimeEntityId( this.entityId );
            Server.getInstance().broadcastPacket( takeItemEntityPacket );
            this.getChunk().setChanged( true );
            this.close();
            player.getInventory().addItem( this.item );
            player.getInventory().sendContents( player );
        }
    }

    @Override
    public boolean canCollideWith( Entity entity ) {
        return false;
    }

    public Item getItem() {
        return this.item.clone();
    }

    public void setItem( Item item ) {
        this.item = item;
    }

    public long getPickupDelay() {
        return this.pickupDelay;
    }

    public void setPickupDelay( long duration, TimeUnit timeUnit ) {
        this.pickupDelay = Server.getInstance().getCurrentTick() + timeUnit.toMillis( duration ) / 50;
    }

    public Player getPlayerHasThrown() {
        return this.playerHasThrown;
    }

    public void setPlayerHasThrown( Player playerHasThrown ) {
        this.playerHasThrown = playerHasThrown;
    }

}

File: src/main/java/org/jukeboxmc/entity/item/EntityItem.kt
package org.jukeboxmc.entity.item

import com.nukkitx.protocol.bedrock.packet.AddItemEntityPacket
import com.nukkitx.protocol.bedrock.packet.TakeItemEntityPacket
import org.jukeboxmc.Server
import org.jukeboxmc.block.BlockType
import org.jukeboxmc.entity.Entity
import org.jukeboxmc.entity.EntityMoveable
import org.jukeboxmc.entity.EntityType
import org.jukeboxmc.event.player.PlayerPickupItemEvent
import org.jukeboxmc.item.Item
import org.jukeboxmc.math.Vector
import org.jukeboxmc.player.Player
import org.jukeboxmc.util.Identifier
import java.util.concurrent.TimeUnit

/**
 * @author LucGamesYT
 * @version 1.0
 */
class EntityItem : EntityMoveable() {
    private var item: Item? = null
    var pickupDelay: Long = 0
        private set
    private var isReset = false
    var playerHasThrown: Player? = null
    override fun update(currentTick: Long) {
        super.update(currentTick)
        if (this.isClosed) {
            return
        }
        if (!this.isImmobile) {
            val blockTypeLayer0: BlockType = this.location.world!!.getBlock(
                this.location.blockX,
                this.boundingBox.maxY.toInt(),
                this.location.blockZ,
                0,
                this.location.dimension,
            ).type
            if (blockTypeLayer0 == BlockType.FLOWING_WATER || blockTypeLayer0 == BlockType.WATER) {
                this.velocity.y = this.velocity.y - gravity * -0.015f
            } else {
                this.velocity.y = this.velocity.y - gravity
            }
            this.checkObstruction(this.location.x, this.location.y, this.location.z)
            this.move(this.velocity)
            var friction: Double = (1 - this.drag).toDouble()
            if (this.isOnGround && (Math.abs(this.velocity.x) > 0.00001 || Math.abs(this.velocity.z) > 0.00001)) {
                friction *= 0.6
            }
            this.velocity.x = (this.velocity.x * friction).toFloat()
            this.velocity.y = this.velocity.y * (1 - this.drag)
            this.velocity.z = (this.velocity.z * friction).toFloat()
            if (this.isOnGround) {
                this.velocity.y = this.velocity.y * -0.5f
            }
            this.updateMovement()
        }
        if (this.isCollided && !isReset && this.velocity.squaredLength() < 0.01f) {
            this.velocity = Vector.zero()
            isReset = true
        }
        if (this.age >= TimeUnit.MINUTES.toMillis(5) / 50 || this.location.y <= -64) {
            this.close()
        }
    }

    override val name: String
        get() = "EntityItem"
    override val width: Float
        get() = 0.25f
    override val height: Float
        get() = 0.25f
    override var gravity: Float = 0.04f
    override val type: EntityType
        get() = EntityType.ITEM
    override val identifier: Identifier
        get() = Identifier.fromString("minecraft:item")

    override fun createSpawnPacket(): AddItemEntityPacket {
        val addItemEntityPacket = AddItemEntityPacket()
        addItemEntityPacket.runtimeEntityId = this.entityId
        addItemEntityPacket.uniqueEntityId = this.entityId
        addItemEntityPacket.itemInHand = item!!.toItemData()
        addItemEntityPacket.position = this.location.toVector3f()
        addItemEntityPacket.motion = this.velocity.toVector3f()
        addItemEntityPacket.metadata.putAll(this.metadata.getEntityDataMap())
        return addItemEntityPacket
    }

    override fun onCollideWithPlayer(player: Player) {
        if (Server.instance.currentTick > pickupDelay && !this.isClosed && !player.isDead) {
            val playerPickupItemEvent = PlayerPickupItemEvent(player, item!!)
            Server.instance.pluginManager.callEvent(playerPickupItemEvent)
            if (playerPickupItemEvent.isCancelled || !player.inventory.canAddItem(playerPickupItemEvent.item)) {
                return
            }
            val takeItemEntityPacket = TakeItemEntityPacket()
            takeItemEntityPacket.runtimeEntityId = player.entityId
            takeItemEntityPacket.itemRuntimeEntityId = this.entityId
            Server.instance.broadcastPacket(takeItemEntityPacket)
            this.close()
            player.inventory.addItem(playerPickupItemEvent.item)
            player.inventory.sendContents(player)
        }
    }

    override fun canCollideWith(entity: Entity?): Boolean {
        return false
    }

    fun getItem(): Item {
        return item!!.clone()
    }

    fun setItem(item: Item?) {
        this.item = item
    }

    fun setPickupDelay(duration: Long, timeUnit: TimeUnit) {
        pickupDelay = Server.instance.currentTick + timeUnit.toMillis(duration) / 50
    }
}
